### PR TITLE
Fix edge gaps

### PR DESCRIPTION
**Pull request by tarmack**
Fixes [Issue: Fix or remove edge_gaps](https://github.com/swaywm/sway/issues/2727) by removing them and allowing negative values for `gaps outer` just the way i3 does it.

While this change allows negative values for the outer gaps to essentially negate `gaps inner` it does not allow to set it lower than the negated `gaps inner` value to prevent windows from ending up outside the container.

This pull request may result in issues with the -1 token value in `workspace_config` struct, if you see this going wrong please give me some pointers on how to go about fixing and testing that.

### CODE REVIEW

**Comment by SirCmpwn** [comment]
Thanks for the patch! Can you associate the email you authored these commits with and your GitHub account so they link up nicely?

File: sway/commands/gaps.c [hunk 1]
```diff
@@ -68,15 +43,19 @@ static struct cmd_results *gaps_set_defaults(int argc, char **argv) {
...
-		config->gaps_inner = amount;
+		config->gaps_inner = (amount >= 0) ? amount : 0;
 	} else {
 		config->gaps_outer = amount;
 	}
+
+    // Prevent negative outer gaps from moving windows out of the container.
+    int overflow;
+    overflow = (config->gaps_inner + config->gaps_outer);
+    if (overflow < 0) {
```
Comment (RyanDwyer) [inline sway/commands/gaps.c]: You could do this without introducing a new variable:

    if (config->gaps_outer < -config->gaps_inner) {
        config->gaps_outer = -config->gaps_inner;
    }

File: sway/commands/gaps.c [hunk 2]
```diff
@@ -68,15 +43,19 @@ static struct cmd_results *gaps_set_defaults(int argc, char **argv) {
...
-	}
-
 	if (inner) {
-		config->gaps_inner = amount;
+		config->gaps_inner = (amount >= 0) ? amount : 0;
 	} else {
 		config->gaps_outer = amount;
 	}
+
+    // Prevent negative outer gaps from moving windows out of the container.
```
Comment (RyanDwyer) [inline sway/commands/gaps.c]: This comment should probably say workspace instead of container.

**Comment by RyanDwyer** [comment]
From the associated issue:

> I noticed the workspace_config struct introduced here: Commit: Rename workspace_outputs to workspace_configs and fix memory leak138d10d). I was wondering if the eventual intention is to have this replace the current data in the sway_workplace struct or if they are supposed to keep living side by side?

The `workspace_config` serves as stored configuration for a workspace that doesn't exist yet. When a workspace is created, it looks for a `workspace_config` for that workspace and copies the settings into the workspace struct. The `workspace_config` struct is then not used, unless the workspace is closed and recreated.

Other feedback:

* The -1 in the `workspace_config` is taking effect. You should initialise it to `INT_MIN`, then check for `INT_MIN` when it is copied into the workspace when the workspace is created.
* When using something like `gaps outer current|all set -10` to change a workspace's outer gaps at runtime, the value is set to 0 (see `configure_gaps` function).
* When using `workspace <ws> gaps outer -10`, a check needs to be done here too. This syntax is intended to be used in configuration files so you can configure a single workspace's gaps before it exists.
* There is still documentation for `edge_gaps`, but you probably didn't find it because it's written as `edge\_gaps`.
* Consider updating the gaps documentation to say that negative outer gaps are allowed.

File: sway/commands/gaps.c (hunk 1 above)
Comment (tarmack) [reply sway/commands/gaps.c]: Done

File: sway/commands/gaps.c (hunk 2 above)
Comment (tarmack) [reply sway/commands/gaps.c]: Done

File: sway/commands/gaps.c [hunk 3]
```diff
@@ -68,15 +43,17 @@ static struct cmd_results *gaps_set_defaults(int argc, char **argv) {
...
-
 	if (inner) {
-		config->gaps_inner = amount;
+		config->gaps_inner = (amount >= 0) ? amount : 0;
 	} else {
 		config->gaps_outer = amount;
 	}
+
+    // Prevent negative outer gaps from moving windows out of the workspace.
+    if (config->gaps_outer < -config->gaps_inner) {
```
Comment (RedSoxFan) [inline sway/commands/gaps.c]: Style: Please use tabs for indentation

File: sway/commands/gaps.c [hunk 4]
```diff
@@ -95,8 +72,12 @@ static void configure_gaps(struct sway_workspace *ws, void *_data) {
 		*prop -= data->amount;
 		break;
 	}
-	if (*prop < 0) {
-		*prop = 0;
+    // Prevent invalid gaps configurations.
+    if (ws->gaps_inner < 0) {
```
Comment (RedSoxFan) [inline sway/commands/gaps.c]: Style: Please use tabs for indentation

File: sway/commands/workspace.c [hunk 5]
```diff
@@ -94,7 +95,16 @@ struct cmd_results *cmd_workspace(int argc, char **argv) {
 			return cmd_results_new(CMD_FAILURE, "workspace gaps",
 					"Expected 'workspace <ws> gaps inner|outer <px>'");
 		}
-		*prop = val >= 0 ? val : 0;
+        *prop = val;
```
Comment (RedSoxFan) [inline sway/commands/workspace.c]: Style: Please use tabs for indentation

File: sway/commands/gaps.c [hunk 6]
```diff
@@ -68,15 +43,17 @@ static struct cmd_results *gaps_set_defaults(int argc, char **argv) {
...
 	if (inner) {
-		config->gaps_inner = amount;
+		config->gaps_inner = (amount >= 0) ? amount : 0;
 	} else {
 		config->gaps_outer = amount;
 	}
+
+	// Prevent negative outer gaps from moving windows out of the workspace.
+	if (config->gaps_outer < -config->gaps_inner) {
+	    config->gaps_outer = -config->gaps_inner;
```
Comment (RedSoxFan) [inline sway/commands/gaps.c]: The second level indentation here is still spaces

File: sway/commands/workspace.c [hunk 7]
```diff
@@ -94,7 +95,16 @@ struct cmd_results *cmd_workspace(int argc, char **argv) {
 			return cmd_results_new(CMD_FAILURE, "workspace gaps",
 					"Expected 'workspace <ws> gaps inner|outer <px>'");
 		}
-		*prop = val >= 0 ? val : 0;
+		*prop = val;
+
+		// Prevent invalid gaps configurations.
+		if (wsc->gaps_inner < 0) {
+		    wsc->gaps_inner = 0;
```
Comment (RedSoxFan) [inline sway/commands/workspace.c]: The third level of indentation here is still spaces

File: sway/commands/workspace.c [hunk 8]
```diff
@@ -94,7 +95,16 @@ struct cmd_results *cmd_workspace(int argc, char **argv) {
...
+		}
+		if (wsc->gaps_outer < -wsc->gaps_inner) {
+		    wsc->gaps_outer = -wsc->gaps_inner;
```
Comment (RedSoxFan) [inline sway/commands/workspace.c]: The third level of indentation here is still spaces

**Comment by tarmack** [comment]
Sorry about the spaces, @RedSoxFan. I fixed my editor but didn't quite manage to fix the code..

**Comment by tarmack** [comment]
I kind of fucked up the squash, I'll split the style fixes out again, hold on..

**Comment by SirCmpwn** [comment]
`git rebase -i HEAD~3` then change everything to `fixup` except the first commit. Then you can `git push -f`

**Comment by tarmack** [comment]
Is it customary for the pull requests on Sway to have just the one final commit? If so, maybe it could be added to the contribution guide.

**Comment by SirCmpwn** [comment]
No, but it is desirable for the commits to be neatly arranged.

**Comment by RedSoxFan** [comment]
Thanks!

**Pull request merged**